### PR TITLE
[EOC-56] fix and improve

### PR DESCRIPTION
**Pull request by alemikolo**

### CODE REVIEW

File: server/controllers/shoppingList.js [hunk 1]
```diff
@@ -157,22 +158,30 @@ const getListData = (req, resp) => {
...
+            .exec()
+            .then(cohort => {
+              if (!cohort || cohort.isArchived) {
+                throw new ResponseError('List data not found.', 404);
+              }
+            });
+        }
+      },
+      err => {
+        if (err.name === 'CastError') {
```
Comment (adamklepacz) [inline server/controllers/shoppingList.js]: can you explain here a little bit?
Wy you checking is err.name is 'CastError' and not just checking if err exist?

File: server/controllers/shoppingList.js [hunk 2]
```diff
@@ -194,13 +203,18 @@ const getListData = (req, resp) => {
...
     })
     .catch(err => {
-      resp.status(400).send({
-        message:
-          'An error occurred while fetching the list data. Please try again.'
-      });
+      if (err instanceof ResponseError) {
+        const { status, message } = err;
+        resp.status(status).send({ message });
+      } else {
```
Comment (adamklepacz) [inline server/controllers/shoppingList.js]: else not necessary here

File: server/controllers/shoppingList.js [hunk 3]
```diff
@@ -194,13 +203,18 @@ const getListData = (req, resp) => {
...
+      resp.status(200).json(data);
     })
     .catch(err => {
-      resp.status(400).send({
-        message:
-          'An error occurred while fetching the list data. Please try again.'
-      });
+      if (err instanceof ResponseError) {
+        const { status, message } = err;
+        resp.status(status).send({ message });
```
Comment (adamklepacz) [inline server/controllers/shoppingList.js]: add return here

File: server/controllers/cohort.js [hunk 4]
```diff
@@ -142,24 +143,30 @@ const getCohortDetails = (req, resp) => {
...
     })
     .catch(err => {
-      resp.status(400).send({
-        message:
-          'An error occurred while deleting the cohort. Please try again.'
-      });
+      if (err instanceof ResponseError) {
+        const { status, message } = err;
+        resp.status(status).send({ message });
+      } else {
```
Comment (adamklepacz) [inline server/controllers/cohort.js]: else not necessary

File: server/controllers/cohort.js [hunk 5]
```diff
@@ -142,24 +143,30 @@ const getCohortDetails = (req, resp) => {
...
         .send({ message: `Cohort ${documentName} successfully deleted.` });
     })
     .catch(err => {
-      resp.status(400).send({
-        message:
-          'An error occurred while deleting the cohort. Please try again.'
-      });
+      if (err instanceof ResponseError) {
+        const { status, message } = err;
+        resp.status(status).send({ message });
```
Comment (adamklepacz) [inline server/controllers/cohort.js]: add return here

File: server/controllers/shoppingList.js (hunk 1 above)
Comment (alemikolo) [reply server/controllers/shoppingList.js]: This code will be removed. It's unclear and not proper. It was added to handle this specific error which occurs when id is not valid (e.g. too short). Before that, in this situation catch method was handling this error, and was returning 400. This solution handle this specific error and return 404 and frontend app can react on it.
We should think about all actions which user can perform and create proper scenariouses of handling errors and when and how show to user proper message.

File: server/common/utilities.js [hunk 6]
```diff
@@ -9,8 +9,10 @@ const filter = f => object =>
   );
 
 const checkRole = (idsArray, userIdFromReq) => {
-  const userId = mongoose.Types.ObjectId(userIdFromReq);
+  const userId = ObjectId(userIdFromReq);
   return idsArray.some(id => id.equals(userId));
 };
 
-module.exports = { checkRole, filter };
+const isIdValid = id => ObjectId.isValid(id);
```
Comment (adamklepacz) [inline server/common/utilities.js]: isValidMonogId

File: server/common/exceptions/HttpException.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class HttpException extends Error {
+  constructor(status, message) {
+    super(message);
+    this.status = status;
+    this.message = message;
```
Comment (adamklepacz) [inline server/common/exceptions/HttpException.js]: I think this not necessary here if you call super(message) from Error class

File: server/controllers/shoppingList.js [hunk 8]
```diff
@@ -194,9 +208,13 @@ const getListData = (req, resp) => {
         ? { _id, cohortId, description, isAdmin, isArchived, name, products }
         : { _id, cohortId, description, isArchived, name, products };
```
Comment (adamklepacz) [inline server/controllers/shoppingList.js]: in previous PR this was diffrent, as I said earlier you can send isAdmin with value false or true and this ternary will be not need

**Pull request merged**